crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,355 @@
+            fetch_tasks,
+        }
+    }
+
+    async fn fetch_task(
+        state: Arc<Mutex<State>>,
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<tokio::sync::Mutex<Receiver<FetchRequest>>>,
+        peer_store: peer_store::DynPeerStore,
+        space_id: SpaceId,
+        transport: DynTransport,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            let is_agent_cooling_down = {
+                let mut state_lock = state.lock().unwrap();
+
+                // Do nothing if op id is no longer in the set of ops to fetch.
+                if !state_lock.ops.contains(&(op_id.clone(), agent_id.clone()))
+                {
+                    continue;
+                }
+
+                state_lock.cool_down_list.is_agent_cooling_down(&agent_id)
+            };
+
+            // Send request if agent is not on cool-down list.
+            if !is_agent_cooling_down {
+                // Send fetch request to agent.
+                let peer = match peer_store.get(agent_id.clone()).await {
+                    Ok(Some(peer)) => match &peer.url {
+                        Some(url) => url.clone(),
+                        None => {
+                            tracing::warn!(
+                                "agent {agent_id} no longer in peer store"
+                            );
+                            continue;
+                        }
+                    },
+                    Ok(None) => {
+                        tracing::warn!(
+                            "could not find agent id {agent_id} in peer store"
+                        );
+                        continue;
+                    }
+                    Err(err) => {
+                        tracing::warn!("could not get agent id {agent_id} from peer store: {err}");
+                        continue;
+                    }
+                };
+
+                let mut data = bytes::BytesMut::new();
+                data.put(op_id.clone().0 .0);
+                data.put(agent_id.clone().0 .0);
+                let data = data.freeze();
+
+                if let Err(err) = transport
+                    .send_module(
+                        peer,
+                        space_id.clone(),
+                        MOD_NAME.to_string(),
+                        data,
+                    )
+                    .await
+                {
+                    tracing::warn!("could not send fetch request for op {op_id} to agent {agent_id}: {err}");
+                    state
+                        .lock()
+                        .unwrap()
+                        .cool_down_list
+                        .add_agent(agent_id.clone());
+                }
+            }
+
+            // Re-insert the fetch request into the queue.
+            if let Err(err) =
+                fetch_request_tx.try_send((op_id.clone(), agent_id.clone()))
```

Comment:
What is the re-drive logic? If we fail to try-send then how do we make sure that ops sitting in the fetch set make it back into the channel? 

I still think this is necessary to avoid deadlocking, but it introduced a new problem too

Reply:
It would make sense to only re-insert an op if the fetch request could be sent successfully. If the agent was unresponsive and then we add it back to the queue, chances are lower that the op will be received in the meantime until we reach the point of this op again in the queue. Whereas if the fetch request was successful, chances are higher that the op will be received while processing other elements in the queue, so that the op will be gone from the ops to fetch list the next we get to this point in the queue.

Reply:
That's got nothing to do with your question. I think all we can do is remove the item from the set if re-insertion fails.

Reply:
That works. Then we don't have state building up the set that can't be reached.

Reply:
Re-insert ops only when agent responsive and
Remove item from ops set when re-insertion fails https://github.com/holochain/kitsune2/pull/35/commits/0e9d9e7bd79590f7ed13d0dbd4b3c5fb026285ca